raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -439,20 +453,28 @@ private void onBecomeFollower() {
 
         // After becoming a follower, we need to complete pending fetches so that
         // they can be resent to the leader without waiting for their expiration
-        purgatory.completeAll(null);
-
-        // TODO: we may only return the future when the append is committed not appended locally,
-        //       in which case we would change this logic
-        uncommittedAppends.clear();
+        fetchPurgatory.completeAll(null);
 
         for (UnwrittenAppend unwrittenAppend: unwrittenAppends) {
-            if (!unwrittenAppend.isCancelled()) {
+            if (!unwrittenAppend.future.isDone()) {
                 unwrittenAppend.fail(new NotLeaderForPartitionException("Append refused since this node is no longer " +
                         "the leader"));
             }
         }
         unwrittenAppends.clear();
 
+        for (UncommittedAppend uncommittedAppend: uncommittedAppends.values()) {
+            if (!uncommittedAppend.future.isDone()) {
+                uncommittedAppend.future.completeExceptionally(new NotLeaderForPartitionException("Append uncommitted " +
+                        "since this node is no longer the leader"));
+            }
+        }
+        uncommittedAppends.clear();
+
+        // Clearing the append purgatory should take no-op other than freeing the objects since they should have all
+        // been completed exceptionally as not-leader
+        appendPurgatory.completeAll(null);
```

Comment:
It seems append purgatory will accumulate requests until they timeout or the leader changes. For a high throughput topic scenario, I guess the collection could get large.

Reply:
Yes. Though that should be relatively the same scale for current produce purgatory right?